### PR TITLE
Copr: add cast_*_as_time

### PR DESCRIPTION
**Pull request by iosmanthus**
###  What have you changed?

Fix rest of cast functions: `cast_*_as_time`

Don't assume reviewers understand the original issue.

###  What is the type of the changes?

Pick one of the following and delete the others:
- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?
Please select the tests that you ran to verify your changes:
- Unit test

### CODE REVIEW

File: components/tidb_query/src/rpn_expr/impl_cast.rs [hunk 1]
```diff
@@ -889,6 +898,158 @@ cast_as_duration!(
...
+        // Parse `val` as a `u64`
+        Time::parse_from_u64(
+            ctx,
+            val as u64,
+            extra.ret_field_type.as_accessor().tp().try_into()?,
+            extra.ret_field_type.get_decimal() as i8,
+        )
+        .map(Some)
+        .or_else(|_| {
+            ctx.handle_invalid_time_error(Error::incorrect_datetime_value(val))
```
Comment (lonng) [inline components/tidb_query/src/rpn_expr/impl_cast.rs]: Why not
```rust
            ctx.handle_invalid_time_error(Error::incorrect_datetime_value(val))
                .map(|_| None)?
```

File: components/tidb_query/src/rpn_expr/impl_cast.rs [hunk 2]
```diff
@@ -889,6 +898,158 @@ cast_as_duration!(
...
+fn cast_int_as_time(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Int>,
+) -> Result<Option<Time>> {
+    if let Some(val) = *val {
+        // Parse `val` as a `u64`
+        Time::parse_from_u64(
+            ctx,
+            val as u64,
```
Comment (lonng) [inline components/tidb_query/src/rpn_expr/impl_cast.rs]: What behavior does MySQL have if the value is a negative number?

File: components/tidb_query/src/rpn_expr/impl_cast.rs (hunk 1 above)
Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_cast.rs]: Done

File: components/tidb_query/src/rpn_expr/impl_cast.rs (hunk 2 above)
Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_cast.rs]: return `NULL`

Comment (lonng) [reply components/tidb_query/src/rpn_expr/impl_cast.rs]: I think we should check the `val` range (There is an error if the `val` less than 101, which includes negative number). See: https://github.com/pingcap/tidb/blob/master/types/time.go#L1278:L1278

Comment (lonng) [reply components/tidb_query/src/rpn_expr/impl_cast.rs]: I think we should check the `val` range (There is an error if the `val` less than 101, which includes negative number). See: https://github.com/pingcap/tidb/blob/master/types/time.go#L1278:L1278

File: components/tidb_query/src/rpn_expr/impl_cast.rs [hunk 3]
```diff
@@ -889,6 +898,138 @@ cast_as_duration!(
...
+fn cast_string_as_time(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra,
+    val: &Option<Bytes>,
+) -> Result<Option<Time>> {
+    if let Some(val) = val {
+        // Convert `val` to a string first and then parse it as a float string.
+        Time::parse(
+            ctx,
+            unsafe { std::str::from_utf8_unchecked(val) },
```
Comment (mapleFU) [inline components/tidb_query/src/rpn_expr/impl_cast.rs]: Is all data from TiDB guaranteed to be utf8?

Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_cast.rs]: If it is not, time parser will report error.

Comment (breeswish) [reply components/tidb_query/src/rpn_expr/impl_cast.rs]: How about accepting byte for this parse function?

Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_cast.rs]: OK. I will create another pull request to fix this.

**Comment by iosmanthus** [comment]
> Is there any testing for `cast_decimal_as_time`?

Sorry, I miss that. This will be fix in the next pull request.

**Pull request merged**